examples/rust-data-loader/rust/src/lib.rs

```diff
@@ -0,0 +1,263 @@
+            .add_channel("/temperature")
+            .message_count(temperature_count as u64);
+        self.temperature_channel_id.replace(temperature_channel.id);
+
+        Ok(init.build())
+    }
+
+    fn create_iter(&self, args: MessageIteratorArgs) -> Result<Self::MessageIterator, Self::Error> {
+        Ok(NDJsonIterator::open(
+            self.rows.clone(),
+            *self.accelerometer_channel_id.borrow(),
+            *self.temperature_channel_id.borrow(),
+            &args,
+        ))
+    }
+
+    fn get_backfill(&self, args: BackfillArgs) -> Result<Vec<Message>, Self::Error> {
+        let accelerometer_channel_id = *self.accelerometer_channel_id.borrow();
+        let temperature_channel_id = *self.temperature_channel_id.borrow();
+        let want_accelerometer = args.channels.contains(&accelerometer_channel_id);
+        let want_temperature = args.channels.contains(&temperature_channel_id);
+
+        let rows = self.rows.borrow();
+        let mut backfill: Vec<Message> = vec![];
+        if want_accelerometer {
+            let option_backfill_accelerometer = rows
+                .iter()
+                .take_while(|row| {
+                    matches![row, Row::Accelerometer(accel) if seconds_to_nanos(accel.time) < args.time]
+                })
+                .last();
+            if let Some(Row::Accelerometer(accel)) = option_backfill_accelerometer {
+                backfill.push(accel.to_message(accelerometer_channel_id));
+            }
+        }
+        if want_temperature {
+            let option_backfill_temperature = rows
+                .iter()
+                .take_while(|row| {
+                    matches![row, Row::Temperature(temperature) if seconds_to_nanos(temperature.time) < args.time]
+                })
+                .last();
+            if let Some(Row::Temperature(temperature)) = option_backfill_temperature {
+                backfill.push(temperature.to_message(accelerometer_channel_id));
+            }
+        }
+        Ok(backfill)
+    }
+}
+
+struct NDJsonIterator {
+    rows: Rc<RefCell<Vec<Row>>>,
+    index: RefCell<usize>,
+    start: u64,
+    end: u64,
+    channels: BTreeSet<u16>,
+    accelerometer_channel_id: u16,
+    temperature_channel_id: u16,
+}
+
+impl NDJsonIterator {
+    fn open(
+        rows: Rc<RefCell<Vec<Row>>>,
+        accelerometer_channel_id: u16,
+        temperature_channel_id: u16,
+        args: &MessageIteratorArgs
+    ) -> Self {
+        Self {
+            rows: rows.clone(),
+            index: 0.into(),
+            start: args.start_time.unwrap_or(0),
+            end: args.end_time.unwrap_or(u64::MAX),
+            channels: args.channels.iter().copied().collect(),
+            accelerometer_channel_id,
+            temperature_channel_id,
+        }
+    }
+}
+
+impl MessageIterator for NDJsonIterator {
+    type Error = anyhow::Error;
+
+    fn next(&self) -> Option<Result<Message, Self::Error>> {
```

Comment:
I wonder if we should make most of these be `&mut self` and wrap everything in `RefCell` up a layer. I don't know why WIT doesn't default to exclusive references but at least with our usage it should be sound.